### PR TITLE
Copy-on-write inside fork syscall

### PR DESCRIPTION
**Pull request by bkjg**
Add copy-on-write to fork.

### CODE REVIEW

**Comment by pj1031999** [comment]
@bkjg what is a status of that PR? Should I start reviewing the code?

**Comment by bkjg** [comment]
> @bkjg what is a status of that PR? Should I start reviewing the code?

@pj1031999 To successful review of this PR we should wait until #831 is merged, because this PR depends on changes from that PR, so they are visible here too.

**Comment by cahirwpz** [comment]
Please resolve conflicts after #831 merge.

**Review by j-piecuch — CHANGES_REQUESTED** [review]

File: sys/kern/vm_physmem.c
Comment (pj1031999) [inline sys/kern/vm_physmem.c:255]: Why `\n`?

File: include/sys/vm.h
Comment (pj1031999) [inline include/sys/vm.h:64]: Maybe some prefix like `vmp_` will be a good idea here.

File: include/sys/vm_object.h
Comment (pj1031999) [inline include/sys/vm_object.h:25]: Maybe some prefix like `vmo_` will be a good idea here.

File: lib/libc/gen/getpagesize.c
Comment (pj1031999) [inline lib/libc/gen/getpagesize.c:1]: `machine/vm_param.h` is already included in `sys/vm.h`.

File: sys/kern/vm_map.c
Comment (pj1031999) [inline sys/kern/vm_map.c:362]: Could you explain what should be done here with pages marked as `VM_PROT_READ | VM_PROT_EXEC`?

File: sys/kern/vm_pager.c
Comment (pj1031999) [inline sys/kern/vm_pager.c:66]: It's hard to read that code.
At the end of that function you always call `vm_object_add_page` so you could write

```
  if (...) {
    ...
    goto out;
  }
  
  ...

out:
  vm_object_add_page(obj, offset, new_pg);
  return new_pg;
```
It is simpler for me.

I also don't understand logic here.
You call 2x `refcnt_acquire` and after that `refcnt_release` – it looks strange without explanation.

**Comment by cahirwpz** [comment]
Please merge `master` into your branch and resolve conflicts.

File: include/sys/vm_object.h
Comment (bkjg) [reply include/sys/vm_object.h]: Great point! Thanks! Changed.

File: sys/aarch64/pmap.c
Comment (bkjg) [reply sys/aarch64/pmap.c]: I didn't think about that before. Thanks! Fixed

File: sys/kern/vm_physmem.c
Comment (bkjg) [reply sys/kern/vm_physmem.c]: My mistake. Sorry. I reversed it.

File: bin/utest/vm_map.c
Comment (bkjg) [reply bin/utest/vm_map.c]: Sounds good. I changed it

File: include/sys/vm_object.h
Comment (bkjg) [reply include/sys/vm_object.h:19]: I thought that in case when we are using the cow, reading access will be more common than the writing access so predictably rwlocks would be faster than mutexes.

Comment (j-piecuch) [reply include/sys/vm_object.h:19]: It is true that with cow reads from `vm_objects` will become more frequent, but in order to benefit performance, read accesses must be _far_ more frequent than write accesses, and read accesses must often occur concurrently. As for the first requirement, that may be the case, although I haven't looked through the code. As for the second, I don't think concurrent read accesses are likely, since we don't support SMP, so concurrent access can occur only when a thread is preempted while holding the lock.
I'm also a little concerned about the fact that `rwlock` uses `sleepq` functions, so there may be starvation and priority inversion issues, since you sometimes acquire a write lock while holding a mutex, e.g. in `vm_page_fault()`.

Comment (bkjg) [reply include/sys/vm_object.h:19]: Okay, I see what you mean. I agree with you. That's right. I also am worried a little about the fact that we don't use rwlocks in mimiker frequently, so they aren't tested well, so we don't know if they work properly for sure. I will try to split this PR to several smaller ones, so rwlocks would be separate change that maybe will be skipped.

File: sys/kern/vm_map.c
Comment (bkjg) [reply sys/kern/vm_map.c:357]: Fixed

File: sys/kern/vm_pager.c
Comment (bkjg) [reply sys/kern/vm_pager.c]: This functionality was moved to another PR.

File: include/sys/vm_object.h
Comment (bkjg) [reply include/sys/vm_object.h]: Do you mean adding to each field?

File: sys/kern/vm_map.c
Comment (bkjg) [reply sys/kern/vm_map.c]: I think in case when sth is executable, then we also should set is as readonly and just disregard the fact that this is VM_PROT_EXEC, don't do anything special in that case

File: lib/libc/gen/getpagesize.c
Comment (bkjg) [reply lib/libc/gen/getpagesize.c]: I remember I had some problems with this header, but I restored the old version

File: include/sys/vm_object.h
Comment (pj1031999) [reply include/sys/vm_object.h]: Yes.

File: sys/kern/vm_map.c
Comment (pj1031999) [reply sys/kern/vm_map.c]: Maybe I don't understand something but if you mark page as RO you can't execute any instruction from that page so how it work's in that case? Do you kill new process in that case or make copy of .text?

File: sys/mips/pmap.c
Comment (j-piecuch) [inline sys/mips/pmap.c:352]: Do we need to hold `pv_list_lock` to call `pmap_pte_write()`?

**Review by j-piecuch — CHANGES_REQUESTED** [review]
There's a significant locking issue.

File: sys/kern/vm_map.c
Comment (bkjg) [reply sys/kern/vm_map.c:343]: oh you're right. I didn't notice it. I will fix it

File: include/sys/vm_map.h
Comment (bkjg) [reply include/sys/vm_map.h:15]: hmm I like the idea. I will change it

File: sys/kern/vm_object.c
Comment (bkjg) [reply sys/kern/vm_object.c:86]: Yes, you're right. I will change it

File: sys/mips/pmap.c
Comment (bkjg) [reply sys/mips/pmap.c]: Hmm no. I will fix it.

File: sys/aarch64/pmap.c
Comment (bkjg) [reply sys/aarch64/pmap.c:542]: Okay, I will create a separate PR with this fix.

File: bin/utest/vm_map.c
Comment (pj1031999) [inline bin/utest/vm_map.c:19]: Please don't add empty lines in random pieces of code.

File: bin/utest/vm_map.c
Comment (pj1031999) [inline bin/utest/vm_map.c:100]: That test doesn't work on AArch64 - you get infinity loop  in pmap. Please fix that or explain what is a source of that problem.

**Review by pj1031999 — CHANGES_REQUESTED** [review]

File: bin/utest/vm_map.c
Comment (bkjg) [reply bin/utest/vm_map.c:100]: I cannot reproduce this problem. Could you provide some example of that case? Or what can I do to observe this situation?

File: bin/utest/vm_map.c
Comment (bkjg) [reply bin/utest/vm_map.c]: Fixed.

File: bin/utest/vm_map.c
Comment (pj1031999) [reply bin/utest/vm_map.c]:
```
git clone https://github.com/bkjg/mimiker
cd mimiker
git checkout copy-on-write
make -j4 BOARD=rpi3
./launch --board rpi3 -d test=user_anon_private_simple
```

Comment (bkjg) [reply bin/utest/vm_map.c]: That was because of the bug with permissions. It's fixed now.

File: include/sys/vm_object.h
Comment (j-piecuch) [inline include/sys/vm_object.h:24]: If I understand the code correctly, `shadow_counter` is the number of shadow objects for which this object is the __backing__ object. In other words, it's the number of objects immediately (i.e. without other objects in between) shadowing this object.

File: sys/kern/vm_map.c
Comment (j-piecuch) [inline sys/kern/vm_map.c:402]: I'm concerned about how this would interact with pages mapped without `VM_PROT_READ`... For instance pages mapped as execute-only (I don't know if it's a thing on `aarch64`), or guard pages with `VM_PROT_NONE`: when cloning such a segment, you would enable read access in the page table, and read access wouldn't cause a page fault.

**Review by j-piecuch — CHANGES_REQUESTED** [review]

File: sys/aarch64/pmap.c
Comment (pj1031999) [inline sys/aarch64/pmap.c:519]: I don't understand that change. Why did you move `vm_page_find` before access checking? I don't see any advantages.
Please don't set `pte` and `pg` at the beginning of that function - it's confusing. You always use only values returned by `pmap_lookup_pte` and `vm_page_find` so _default_ 0 is not good here.

**Review by pj1031999 — CHANGES_REQUESTED** [review]

File: sys/kern/vm_object.c
Comment (bkjg) [reply sys/kern/vm_object.c]: Done.

File: sys/mips/pmap.c
Comment (bkjg) [reply sys/mips/pmap.c:479]: I forgot it in `aarch64`. I added it.

Comment (bkjg) [reply sys/mips/pmap.c:479]: The logic is that if I didn't check if the page belongs to backing object, this function returned this EACCES and in the result SIG SEGV was raised. So to fix it, I had to add checking if owner of this page is backing object and if so then we have to call vm_page_fault.

File: include/sys/vm_object.h
Comment (bkjg) [reply include/sys/vm_object.h]: Oh right. My bad. I fixed it.

File: sys/aarch64/pmap.c
Comment (bkjg) [reply sys/aarch64/pmap.c]: I moved `vm_page_find` before access checking because in copy-on-write was the case where we checked access and raised segfault after that. However because the page we checked access for, belonged to the backing object, we shouldn't return `EACCES` but` EFAULT` to make sure `vm_page_fault` will be executed. That's the reason for change.

Comment (pj1031999) [reply sys/aarch64/pmap.c]: But here you don't check if page belongs to backing object.

Comment (bkjg) [reply sys/aarch64/pmap.c]: I forgot to but I fixed it and now I check it.

File: sys/mips/pmap.c
Comment (j-piecuch) [reply sys/mips/pmap.c:479]: I think we need to return `EFAULT` only if the thread attempted to write to a page in a backing object. We could merge that logic into the access check:
```c
if ((prot & VM_PROT_WRITE) && !(pte & PTE_SW_WRITE))
  return (vm_object_is_backing(pg->object) ? EFAULT : EACCES);
```

File: sys/kern/vm_map.c
Comment (bkjg) [reply sys/kern/vm_map.c]: Okay, I added checking if segment this object originally belonged to, has `VM_PROT_READ` set and if so, then I set all pages as readonly.

**Comment by cahirwpz** [comment]
@franciscozdo Please evaluate this PR for useful changes in the context of adapting Cranor's `amap`.

**Comment by franciscozdo** [comment]
Copy-on-write will be handled in another way.

**Comment by cahirwpz** [comment]
Hence we're going to handle CoW in a different way I'm closing this PR.

**Pull request closed without merging**
